### PR TITLE
Let ImageLoadingTransformer dispose the last image it loads

### LINKED ISSUE #4126: Re-using the same Dataview with Bitmaps in memory, breaks when fitting different models or run cross validation on it
### System information

- **OS version/distro**: Windows 10
- **.NET Version (eg., dotnet --info)**: .NET Core 2.2 

### Issue

- **What did you do?**
I had a working pipeline for training image classification with cross-validation on the previous ML.NET version, using file paths as input. Now, being able to load Bitmaps, I am trying to setup a similar pipeline, but allowing training and predictions from in-memory bitmaps.
- **What happened?**
The training works if I just Fit the data,
`ITransformer mlModel = pipeline.Fit(trainData);`
 but it fails if I try to use CrossValidate
`var cvResults = _mlContext.MulticlassClassification.CrossValidate(trainData, pipeline, numberOfFolds);`
- **What did you expect?**
I expected a pipeline that worked with Fit to work with CrossValidate, but it seems the internal multiple passes do something to the Bitmaps (they lose data).
### Source code / logs
My current pipeline, based on [this sample](https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/end-to-end-apps/DeepLearning_ImageClassification_TensorFlow) is this:
```C#
var pipeline = _mlContext.Transforms.Conversion.MapValueToKey("Label")               
                .Append(_mlContext.Transforms.ResizeImages(outputColumnName: TensorFlowModelSettings.inputTensorName, imageWidth: ImageSettings.imageWidth, imageHeight: ImageSettings.imageHeight, inputColumnName: nameof(ImageInputData.Image)))                
                .Append(_mlContext.Transforms.ExtractPixels(outputColumnName: TensorFlowModelSettings.inputTensorName, interleavePixelColors: ImageSettings.channelsLast, offsetImage: ImageSettings.mean/*, inputColumnName: nameof(ImageInputData.Image)*/))                
                .Append(_mlContext.Model.LoadTensorFlowModel(tensorFlowModelFilePath).
                ScoreTensorFlowModel(outputColumnNames: new[] { TensorFlowModelSettings.outputTensorName },
                                    inputColumnNames: new[] { TensorFlowModelSettings.inputTensorName }, addBatchDimensionInput: false))                
                .Append(_mlContext.MulticlassClassification.Trainers.LbfgsMaximumEntropy(labelColumnName: "Label", featureColumnName: TensorFlowModelSettings.outputTensorName))
                .Append(_mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel"))
                .AppendCacheCheckpoint(_mlContext);
```

The error log includes the following exceptions:

```
System.ArgumentException: Parameter is not valid.
   at System.Drawing.Image.get_Height()
   at Microsoft.ML.Transforms.Image.ImageResizingTransformer.Mapper.<>c__DisplayClass3_0.<MakeGetter>b__1(Bitmap& dst)
   at Microsoft.ML.Transforms.Image.ImagePixelExtractingTransformer.Mapper.<>c__DisplayClass5_0`1.<GetGetterCore>b__1(VBuffer`1& dst)
   at Microsoft.ML.Data.DataViewUtils.Splitter.InPipe.Impl`1.Fill()
   at Microsoft.ML.Data.DataViewUtils.Splitter.<>c__DisplayClass9_0.<SplitCore>b__1()
```

```
System.InvalidOperationException: Splitter/consolidator worker encountered exception while consuming source data ---> System.ArgumentException: Parameter is not valid.
   at System.Drawing.Image.get_Height()
   at Microsoft.ML.Transforms.Image.ImageResizingTransformer.Mapper.<>c__DisplayClass3_0.<MakeGetter>b__1(Bitmap& dst)
   at Microsoft.ML.Transforms.Image.ImagePixelExtractingTransformer.Mapper.<>c__DisplayClass5_0`1.<GetGetterCore>b__1(VBuffer`1& dst)
   at Microsoft.ML.Data.DataViewUtils.Splitter.InPipe.Impl`1.Fill()
   at Microsoft.ML.Data.DataViewUtils.Splitter.<>c__DisplayClass9_0.<SplitCore>b__1()
   --- End of inner exception stack trace ---
   at Microsoft.ML.Data.DataViewUtils.Splitter.Batch.SetAll(OutPipe[] pipes)
   at Microsoft.ML.Data.DataViewUtils.Splitter.Cursor.MoveNextCore()
   at Microsoft.ML.Data.RootCursorBase.MoveNext()
   at Microsoft.ML.Data.DataViewUtils.Splitter.<>c__DisplayClass5_1.<ConsolidateCore>b__2()
```

This is my first issue here, and I apologize if I overlooked something. I found no posts about this error anywhere.

### PR DESCRIPTION
**Pull request by antoniovs1029**
Fixes #4126.

By investigating issue #4126 I found out that `ImageResizingTransformer `disposed the last input Bitmap object it resized (i.e. _only_ the last image object in the input dataset). When working with in-memory images (i.e. NOT loaded with `ImageLoaderTransformer`) this is inadequate, since the `ImageResizingTransformer `will dispose the last image in the dataset, and the user won't be able to use that image again. Particullarly, issue #4126 happened when running Cross Validation with a pipeline that included applying `ImageResizingTransformer `to in memory images: when scoring the first model during cross validation, the last image of the input dataset got disposed, and when trying to score the second model, the image wasn't there, and a exception was thrown.

The only reason I found as to why `ImageResizingTransformer` had this behavior, is because `ImageLoadingTransformer` actually disposes all the images it loads except for the last one ([link](https://github.com/dotnet/machinelearning/blob/919bc8b886a6bc20c5022e8dad00f84bc1248236/src/Microsoft.ML.ImageAnalytics/ImageLoader.cs#L226-L230)). So assuming this is the only reason for `ImageResizingTransformer's `disposer, I simply moved the disposal mechanism of the last image to `ImageLoadingTransformer`.

I added two tests that used to throw exceptions without the changes on this PR. And I also manually tested with the debugger's profiler that the last image got correctly disposed when using `ImageLoadingTransformer`.

### CODE REVIEW

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 1]
```diff
@@ -1126,6 +1127,81 @@ public void TensorFlowTransformCifarSavedModel()
             }
         }
 
+        public class InMemoryImage
+        {
```
Comment (harishsk) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1131]: Is this the same class as the one in ImageTests.cs?
They both seem to be in the same project. Can one of them be eliminated?

Comment (antoniovs1029) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1131]: Yeah, they're the same class. But I wouldn't know how to share them across those tests without messing up the namespaces (one test is in Microsoft.ML.Tests and the other one in Microsoft.ML.Scenarios), so I just think it's better to keep them separate.

Comment (harishsk) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1131]: You can simply add a using Microsoft.ML.Tests in the this file and reference the class.

---
In reply to: [414190280](https://github.com/dotnet/machinelearning/pull/5056#discussion_r414190280) [](ancestors = 414190280)

Comment (antoniovs1029) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1131]: Yes, but that would make accessible anything from Microsoft.ML.Tests on Microsoft.ML.Scenarios, and that wouldn't be right?

Comment (antoniovs1029) [reply test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:1131]: Ok, I'll use "using InMemoryImage = Microsoft.ML.Tests.ImageTests.InMemoryImage;" inside the tensorflow tests :wink:

**Pull request merged**